docs/proposals/device-crd.md

```diff
@@ -0,0 +1,556 @@
+
+## Motivation
+
+Device management is a key feature required for IoT use-cases in edge computing.
+This proposal addresses how can we manage devices from the cloud and synchronize
+the device updates between edge nodes and cloud.
+
+### Goals
+
+Device management must:
+* provide APIs for managing devices from the cloud.
+* synchronize the device updates between cloud and edge nodes.
+
+### Non-goals
+
+* To design **secure** device provisioning.
+* To address OTA device firmware upgrades.
+* To address how device auto-discovery can happen.
+* To address device migration scenarios.
+
+## Proposal
+We propose using Kubernetes [Custom Resource Definitions (CRDs)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) to describe device metadata/status and a controller to synchronize these device updates between edge and cloud.
+<img src="../images/device-crd/device-crd-model.png">
+
+### Use Cases
+
+* Describe device properties.
+  * Users can describe device proeprties and access mechanisms to interact with / control the device.
+* Perform CRUD operations on devices from cloud.
+   * Users can create, update and delete device metadata from the cloud via the CRD APIs exposed by the Kubernetes API server.
+   * Users can control the expected state of a device via the CRD APIs.
+* Report device status from the edge node.
+  * Mapper applications running on the edge can report the device state as well the actual values of the control attributes.
+
+## Design Details
+
+### Device model CRD
+<img src="../images/device-crd/device-model-crd.png">
+
+A `device model` describes the device properties exposed by the device and property visitors to access these properties. A device model is like a reusable template using which many devices can be created and managed.
+
+### Device Model Type Definition
+```go
+// DeviceModelSpec defines the model / template for a device.It is a blueprint which describes the device
+// capabilities and access mechanism via property visitors.
+type DeviceModelSpec struct {
+  // List of device attributes.
+  // Required.
+  DeviceProperties       []DeviceProperty        `json:"properties,omitempty"`
+  // List of property visitors which describe how to access the device attributes.
+  // Required.
+  DevicePropertyVisitors []DevicePropertyVisitor `json:"propertyVisitors,omitempty"`
+}
+
+// DeviceProperty describes an individual device property / attribute like temperature / humidity etc.
+type DeviceProperty struct {
+  // The property name.
+  // Required.
+  Name        string            `json:"name,omitempty"`
+  // The property description.
+  // +optional
+  Description string            `json:"description,omitempty"`
+  // Additional metadata about the device property like datatype, range, min, max etc.
+  // Optional: If no metadata is provided, it is assumed that the property is an empty read-only string.
+  // +optional
+  Metadata    map[string]string `json:"metadata,omitempty"`
```

Comment:
A silly question, who will update the `Metadata` field?

Reply:
these could be updated by the user who creates the device model via kubectl. The update would be a kubectl patch operation.